design/dsl/api.go

```diff
@@ -269,149 +270,64 @@ func ResponseTemplate(name string, p interface{}) {
-				Name:     name,
-				Template: t,
-			}
-		} else if tmpl, ok := p.(func(v1, v2, v3, v4, v5, v6 string)); ok {
-			t := func(params ...string) *design.ResponseDefinition {
-				if len(params) < 6 {
-					ReportError("expected six arguments when invoking response template %s", name)
-					return nil
-				}
-				r := &design.ResponseDefinition{Name: name}
-				executeDSL(func() { tmpl(params[0], params[1], params[2], params[3], params[4], params[5]) }, r)
-				return r
-			}
-			a.ResponseTemplates[name] = &design.ResponseTemplateDefinition{
-				Name:     name,
-				Template: t,
-			}
-		} else if tmpl, ok := p.(func(v1, v2, v3, v4, v5, v6, v7 string)); ok {
-			t := func(params ...string) *design.ResponseDefinition {
-				if len(params) < 7 {
-					ReportError("expected seven arguments when invoking response template %s", name)
-					return nil
-				}
-				r := &design.ResponseDefinition{Name: name}
-				executeDSL(func() { tmpl(params[0], params[1], params[2], params[3], params[4], params[5], params[6]) }, r)
-				return r
-			}
-			a.ResponseTemplates[name] = &design.ResponseTemplateDefinition{
-				Name:     name,
-				Template: t,
-			}
-		} else if tmpl, ok := p.(func(v1, v2, v3, v4, v5, v6, v7, v8 string)); ok {
-			t := func(params ...string) *design.ResponseDefinition {
-				if len(params) < 8 {
-					ReportError("expected eight arguments when invoking response template %s", name)
-					return nil
-				}
-				r := &design.ResponseDefinition{Name: name}
-				executeDSL(func() { tmpl(params[0], params[1], params[2], params[3], params[4], params[5], params[6], params[7]) }, r)
-				return r
-			}
-			a.ResponseTemplates[name] = &design.ResponseTemplateDefinition{
-				Name:     name,
-				Template: t,
+			executeDSL(func() { tmpl(params...) }, r)
+			return r
+		}
+		a.ResponseTemplates[name] = &design.ResponseTemplateDefinition{
+			Name:     name,
+			Template: t,
+		}
+	} else {
+		typ := reflect.TypeOf(p)
+		if kind := typ.Kind(); kind != reflect.Func {
+			ReportError("dsl should be a function but get %v", kind)
+			return
+		}
+
+		if num := typ.NumOut(); num > 0 {
+			ReportError("dsl should have 0 return parameter but get %d", num)
+			return
+		}
+
+		num := typ.NumIn()
+		val := reflect.ValueOf(p)
+		t := func(params ...string) *design.ResponseDefinition {
+			if len(params) < num {
+				ReportError("expected at least %d argument when invoking response template %s", num, name)
+				return nil
 			}
-		} else if tmpl, ok := p.(func(v1, v2, v3, v4, v5, v6, v7, v8, v9 string)); ok {
-			t := func(params ...string) *design.ResponseDefinition {
-				if len(params) < 9 {
-					ReportError("expected nine arguments when invoking response template %s", name)
+			r := &design.ResponseDefinition{Name: name}
+
+			in := make([]reflect.Value, num)
+			for i := 0; i < num; i++ {
+				// type checking
+				if t := typ.In(i); t.Kind() != reflect.String {
+					ReportError("expect parameter type string but get %s", t)
```

Comment:
Please use:
`("ResponseTemplate parameters must be strings but type of parameter at position %d is %s", i, t)`